Optimized Min, Max and  SequenceEqual.

akeit0 opened the pull request:
Min, Max, SequenceEqualを最適化しました。
Intrinsicsは消しつつ、AVX2環境で
int Min Avg 0.76 μs ->0.24 μs
Int  SequenceEqual 5.72 μs ->2.37 μs
double Min Avg 18.3 μs ->4.0 μs
となっています。
byte型に関してはすべて10倍ほど高速化しています。

akeit0 committed:
Optimize MinMax by
AVX support and
introducing local functions to make it easier for the compiler to vectorize.
MinMax of double is  4 times faster.
MinMax of byte is  10 times faster.
Not tested but minmax for float, sbyte, short, ushort, long and ulong is  supposed to be faster.

akeit0 committed:
Optimized MinMax
Removed intrinsics, and made more vectorized code.
Optimized SequenceEqual
Classified into integer, single precision, and double precision, and optimized for each

Int SequenceEqual

Min:		5.50 μs
Median:		5.70 μs
Max:		6.70 μs
Avg:		5.72 μs
StdDev:		0.22 μs
SampleCount:	100
Sum:		571.60 μs

Optimized

Min:		2.20 μs
Median:		2.30 μs
Max:		4.60 μs
Avg:		2.37 μs
StdDev:		0.33 μs
SampleCount:	100
Sum:		237.10 μs

Int Min:
Min:		0.70 μs
Median:		0.80 μs
Max:		0.90 μs
Avg:		0.76 μs
StdDev:		0.05 μs
SampleCount:	100
Sum:		76.00 μs

Optimized

Min:		0.10 μs
Median:		0.20 μs
Max:		1.40 μs
Avg:		0.24 μs
StdDev:		0.17 μs
SampleCount:	100
Sum:		24.00 μs

AnnulusGames commented:
ありがとうございます。
こちらでプロジェクトを開いたところ'MinCore'が見つからない旨のコンパイルエラーが発生します。Minがコミットに含まれていないのが原因かと思われるのですが...

AnnulusGames commented:
もう一つ、intのSequenceEqualの試行回数を増やすと単体テストを通過しないことがあります。詳しくはまだ調べられていませんが、特定の状況下で通常のLINQと異なる結果を返す可能性があります。

Akeit0 commented:
すみません、調べてみます。

Akeit0 commented:
上手く行かなかったのはアライメントの問題でした。Unsafe.ReadUnalignedか同等のilが必要ですね

AnnulusGames commented:
Unsafeクラスのdllの追加は他ライブラリとの競合の可能性があるため、Unsafe.ReadUnalignedを使用することは難しいです。UnsafeUtility.ReadArrayElement<T>を用いることで同等の処理が可能でしょうか...?

Akeit0 commented:
ReadArrayElementはunalignedではないので、IL書くしかないですね。
Unity上で作るなら、Mono.Cecilで書き換える感じです。

AnnulusGames commented:
Mono.Cecilの利用は安定性やメンテナンス性の面から可能な限り避けたいと考えています。
現状Min/Maxは問題なく動作するようなので、ひとまずそれら2つをPRに含めるのが良いかと思うのですがいかがでしょうか。

Akeit0 commented:
Unsafe.ReadUnalignedがなくてもとりあえず大丈夫そうです。

akeit0 committed:
Added bool8, bool16, bool32.
Fixed the memory alignment issue.
Changed to branch by supported Vector width.

akeit0 committed:
Added bool8, bool16, bool32.
Fixed the memory alignment issue.
Changed to branch by supported Vector width.
Made more vectorization to Sums of some vector types.
Vectorized more explicitly.

akeit0 committed:
Fixed the push conflict issue.
See the previous push.(About bools)

akeit0 committed:
Fixed ContainsCore bug.
Add some tests for vector types.

akeit0 committed:
Remove unnecessary codes.

Akeit0 commented:
問題解消しました。
Average以外ほぼすべてのCore関数をサポートされるVectorの長さで分岐するようにしました。
テストもVector2用のものを追加するなどした上でうまく行っています。

AnnulusGames commented:
こちらでも問題なく動作することを確認したのでマージしました。ありがとうございます...！
